__Name:__ Joaquin Quintana

In [1]:
import numpy as np

1. 

* __The 4 properties of matrix multiplication.__

* __The inner product and matrix-vector products__ pg.179

* __Inner product representation__ 

* __A block matrix multiplication__

* __A column interpretation of matrix-matrix product__

* __Multiple sets of linear equations (skip row interpretation if you wish)__

* __Gram Matrix__

* __Outer Product representation__

## The 4 properties of matrix multiplication.

1. Associativity: $(AB)C = A(BC)$

2. Associativity with scalar multiplication: $\alpha(AB)= (\alpha A)B$

3. Distributivity with addition: $A(B+C) = (AB) + (AC) and A+B(C) = (AC) + (BC)$

4. Transpose of product: $(AB)^T = B^TA^T$

The four code blocks below demonstrates each of these properties

In [2]:
#Associativity:  (𝐴𝐵)𝐶=𝐴(𝐵𝐶)
A = np.random.rand(3,3)
B = np.random.rand(3,3)
C = np.random.rand(3,3)
print("(AB)C = \n", (A@B)@C,"\n")
print("A(BC) = \n", A@(B@C),"\n")
#There seems to be some rounding error here so I simply dropped some trailing digits to show they are close. 
print("(AB)C == A(BC)\n", np.round( (A@B)@C,5) == np.round( A@(B@C),5) )

(AB)C = 
 [[1.06968138 0.78943946 0.93637945]
 [1.6376566  1.29639002 1.57167332]
 [1.44686589 1.13288483 1.34080172]] 

A(BC) = 
 [[1.06968138 0.78943946 0.93637945]
 [1.6376566  1.29639002 1.57167332]
 [1.44686589 1.13288483 1.34080172]] 

(AB)C == A(BC)
 [[ True  True  True]
 [ True  True  True]
 [ True  True  True]]


In [3]:
# Associativity with scalar multiplication:  𝛼(𝐴𝐵)=(𝛼𝐴)𝐵
#let alpha be some random scalar 
alpha = np.random.rand(1,1)
#A and B are matricies
A = np.random.rand(3,3)
B = np.random.rand(3,3)

print("𝛼(𝐴𝐵) = \n", alpha*(A@B))
print("\n(𝛼𝐴)𝐵 = \n", (alpha*A)@B)
print("\n Are 𝛼(𝐴𝐵)=(𝛼𝐴)𝐵: \n", np.round( alpha*(A@B),5) == np.round( (alpha*A)@B ,5) )

𝛼(𝐴𝐵) = 
 [[0.23851754 0.13721401 0.12313238]
 [0.51358982 0.32869094 0.26206031]
 [0.28627294 0.14272212 0.13456416]]

(𝛼𝐴)𝐵 = 
 [[0.23851754 0.13721401 0.12313238]
 [0.51358982 0.32869094 0.26206031]
 [0.28627294 0.14272212 0.13456416]]

 Are 𝛼(𝐴𝐵)=(𝛼𝐴)𝐵: 
 [[ True  True  True]
 [ True  True  True]
 [ True  True  True]]


In [4]:
#Distributivity with addition:  𝐴(𝐵+𝐶)=(𝐴𝐵)+(𝐴𝐶)𝑎𝑛𝑑 𝐴+𝐵(𝐶)=(𝐴𝐶)+(𝐵𝐶)
A = np.random.rand(3,3)
B = np.random.rand(3,3)
C = np.random.rand(3,3)
print("𝐴(𝐵+𝐶) = \n", A@(B+C))
print("\n(𝐴𝐵)+(𝐴𝐶) = \n", (A@B)+(A@C))
print("\n(Is 𝐴(𝐵+𝐶)=(𝐴𝐵)+(𝐴𝐶) equivalent: \n", np.round (A@(B+C) ,5) == np.round ( (A@B)+(A@C),5) )

print("\n 𝐴+𝐵(𝐶) = \n", (A+B)@C )
print("\n (𝐴𝐶)+(𝐵𝐶) = \n", (A@C)+(B@C) )
print("\n Is 𝐴+𝐵(𝐶)=(𝐴𝐶)+(𝐵𝐶)  equivalent: \n", np.round((A+B)@C,5) == np.round((A@C)+(B@C),5))

𝐴(𝐵+𝐶) = 
 [[0.79787029 1.25036695 0.84969945]
 [0.87630883 1.12653972 1.32046106]
 [0.79705492 1.34518813 1.17361269]]

(𝐴𝐵)+(𝐴𝐶) = 
 [[0.79787029 1.25036695 0.84969945]
 [0.87630883 1.12653972 1.32046106]
 [0.79705492 1.34518813 1.17361269]]

(Is 𝐴(𝐵+𝐶)=(𝐴𝐵)+(𝐴𝐶) equivalent: 
 [[ True  True  True]
 [ True  True  True]
 [ True  True  True]]

 𝐴+𝐵(𝐶) = 
 [[0.9064003  1.63499558 0.70608175]
 [0.94907602 2.09435563 0.96681752]
 [1.63936637 2.49680369 1.05113032]]

 (𝐴𝐶)+(𝐵𝐶) = 
 [[0.9064003  1.63499558 0.70608175]
 [0.94907602 2.09435563 0.96681752]
 [1.63936637 2.49680369 1.05113032]]

 Is 𝐴+𝐵(𝐶)=(𝐴𝐶)+(𝐵𝐶)  equivalent: 
 [[ True  True  True]
 [ True  True  True]
 [ True  True  True]]


In [5]:
#Transpose of product:  (𝐴𝐵)^𝑇=𝐵^𝑇 𝐴^𝑇
#using the same matricies from above 
#compute the transpose of AB
print("(AB)^T = \n", (A@B).T)
print("\n B^T A^T = \n", B.T@A.T)

(AB)^T = 
 [[0.15136657 0.51947227 0.2874851 ]
 [0.43506092 0.31494006 0.44530844]
 [0.53397323 0.94471086 0.7883784 ]]

 B^T A^T = 
 [[0.15136657 0.51947227 0.2874851 ]
 [0.43506092 0.31494006 0.44530844]
 [0.53397323 0.94471086 0.7883784 ]]


# inner product and matrix-vector products

Let $y$ be an $m$-vector, $x$ be an $n$-vector and $A$ an $m*n$ matrix. We will show that $y^T(Ax) = (y^TA)x = (A^Ty)^Tx$. In python...

In [8]:
#creat A
A = np.random.rand(6,3)
m,n = A.shape
#create y 
y = np.random.rand(m)
#create x
x = np.random.rand(n)

compute $y^T(Ax)$

In [9]:
y.T@(A@x)

2.908483497699705

compute $(y^TA)x$

In [10]:
(y.T@A)@x

2.9084834976997045

compute $(A^Ty)^Tx$

In [11]:
(A.T@y).T@x

2.9084834976997045

Show the equivalence between these outputs. This, $y^T(Ax) == (y^TA)x == (A^Ty)^Tx$

In [12]:
np.round(y.T@(A@x),5) == np.round((y.T@A)@x,5) == np.round((A.T@y).T@x,5)

True

# Block Matrix

Here let A and B block matricies of size $4*4$. Each block matrix is compose of four matricies of the same size denoted with lowercase letter. Following from formula 10.1 we can let C = AB to get. 

$C = AB =  \begin{bmatrix} a & b\\ c & d \end{bmatrix} \begin{bmatrix} e & f\\ g & h \end{bmatrix} = \begin{bmatrix} ae + bg & af + bh \\ ce +dg & cf +dh \end{bmatrix} $

In [13]:
#create sub matricies pg.180
a = np.random.rand(2,2)
b = np.random.rand(2,2)
c = np.random.rand(2,2)
d = np.random.rand(2,2)
e = np.random.rand(2,2)
f = np.random.rand(2,2)
g = np.random.rand(2,2)
h = np.random.rand(2,2)
#create block matricies A and B
A= np.block([[a,b],[c,d]])
B= np.block([[e,f],[g,h]])

__Show C is equivalent to A@B__: (note if checking equivalence some rounding error will exist between the two results).

In [14]:
A@B

array([[0.97404665, 0.93076371, 1.18026678, 1.64005916],
       [0.69031591, 0.68402325, 0.83329102, 1.12355135],
       [0.15418921, 0.11313369, 0.28278045, 0.34891251],
       [0.37400751, 0.43837581, 0.58286326, 0.6600701 ]])

In [15]:
C = np.block([[a@e + b@g, a@f + b@h],[c@e + d@g, c@f + d@h]])
C

array([[0.97404665, 0.93076371, 1.18026678, 1.64005916],
       [0.69031591, 0.68402325, 0.83329102, 1.12355135],
       [0.15418921, 0.11313369, 0.28278045, 0.34891251],
       [0.37400751, 0.43837581, 0.58286326, 0.6600701 ]])

# Another block matrix multiplication

In [16]:
A = np.ones((2, 3), int)
B = 9 * A
BlockA = np.block([[A, B],[B,A]]) 
print("BlockA shape: ",BlockA.shape)

BlockA shape:  (4, 6)


In [17]:
D = np.ones((3, 2), int)
Dstack = np.vstack([D,D])
BlockB = np.block([[D,-2*B.T],[D*5,B.T]])
print("BlockB shape: ",BlockB.shape)

BlockB shape:  (6, 4)


In [18]:
C = BlockA@BlockB
C

array([[ 138,  138,  189,  189],
       [ 138,  138,  189,  189],
       [  42,   42, -459, -459],
       [  42,   42, -459, -459]])

# Column interpretation matrix matrix product pg 190
A has size n*p and B p*n

$AB = A[b_1,b_2,...,b_n] = [Ab_1,Ab_2,...,Ab_n]$

In python ...

In [20]:
#make some arrays
A = np.array([[-1.5,3,2],[1,-1,0]])
B = np.array([[-1,-1],[0,-2],[1,0]])
#pg. 181 Inner product representation 
AB1=A@B[:,0]
AB2=A@B[:,1]
#place it all back together 
np.vstack((AB1,AB2)).T

array([[ 3.5, -4.5],
       [-1. ,  1. ]])

## Multiple sets of linear equations 

Let $A$ be a matrix which holds the coefficients for a set of $k$ linear equation with size $m,n$. The k linear equations here must, by the nature of the problem, all have the same $m*n$ coefficent matrix $A$. The linear equation for this is $AX = B$ aka a linear equation with matrix right-hand side. Here $X$ and $B$ are $n*k$ matricies. This equation expanded out is $Ax_i = b_i$ where $i=1,...,k$. An attempt in python with no real meaning but for practice. 

In [19]:
#create A and let the size be m*n
A = np.random.rand(5,5)
m,n = A.shape
k=3

#create X which is a set of variables X = [x_i,...,x_k] each with n rows. 
X = np.random.rand(n,k)#10 linear eq
#compute B
B = [A@X[:,i] for i in range(k)]
#place it all back together 
B = np.vstack((B)).T
#check the shape for B and X are the same 
print("the shape of B: n,k = ", B.shape,"\n","the shape of X: n,k = ", X.shape, "\nB = \n", B)

the shape of B: n,k =  (5, 3) 
 the shape of X: n,k =  (5, 3) 
B = 
 [[1.12466909 1.4500883  0.81870623]
 [1.3538051  1.29675522 0.94656111]
 [1.01124565 1.15532481 0.58316572]
 [1.11966085 0.98524019 0.88635921]
 [0.90829493 1.75110538 0.93341083]]


## The inner product representation

In [34]:
#pg.177
A = np.array([[-1.5,3,2],[1,-1,0]])
B = np.array([[-1,-1],[0,-2],[1,0]])
#pg. 191 Inner product representation. matrix multiplication A row wise and columns B
AB1=A[0,:]@B[:,0]
AB2=A[0,:]@B[:,1]
AB3 = A[1,:]@B[:,0]
AB4 = A[1,:]@B[:,1]

#place it all back together 
np.array([[AB1,AB2],[AB3,AB4]])

array([[ 3.5, -4.5],
       [-1. ,  1. ]])

# Gram Matrix

The gram matrix is created by taking the inner product of the matrix transposed and the same matrix untransposed. This is, $G = A^TA$ transpose ^T. Here is an example in python.

In [21]:
#imples the matrix is transposed first
G = A.T@A
G

array([[ 3.25, -5.5 , -3.  ],
       [-5.5 , 10.  ,  6.  ],
       [-3.  ,  6.  ,  4.  ]])

# Outer Product representation 
Let A be of size $m*p$ and lets express this matrix in terms of it columns which is, $A = [a_1,...,a_p]$. Now let B be a matrix of size $p*n$ and express this matrix B as the rows transposed which is, $B = \begin{bmatrix} b^T_1 \\.\\.\\.\\ b^T_p\end{bmatrix}$. In python this looks like...

In [22]:
A = np.array([[-1.5,3,2],[1,-1,0]])
B = np.array([[-1,-1],[0,-2],[1,0]])

#This is correctly computing the Outer product representation. 
#make some arrays
m,p = A.shape
out = []
for i in range(p):
    #take the outer product of the columns of a and the rows of b. Make sure b_i has the correct shape here. 
    out.append(np.outer(A[:,i],B[i,:].reshape(2,1)))
          
sum(out)

array([[ 3.5, -4.5],
       [-1. ,  1. ]])

# 10.39


In this problem we see that the gram matrix for A is the same as the gram matrix for R. This lets us know the angle for the columns of R are equal to the angle of the columns of A. Another way of looking at this is from the point of view of the gram-schmidt algorithm. Since gram-schmidt is run on the columns of A and we create R from the result of this output we would expect the relationship between the two matricies to exist as gram-schmidt returns a orthonormalized set of vectors.  

Below I Show that $G = A^TA = (QR)^T(RQ) = R^TQ^TQR = R^T(Q^TQ) = R^TR$ using python.

In [23]:
import numpy as np 

def gram_schmidt(a):
    q = []
    for i in range(len(a)):
    #orthogonalization
        q_tilde = a[i]
        for j in range(len(q)):
            q_tilde = q_tilde - (q[j] @ a[i])*q[j]
        #Test for dependennce
        if np.sqrt(sum(q_tilde**2)) <= 1e-10:
            print('Vectors are linearly dependent.')
            print('GS algorithm terminates at iteration ', i+1)
            return q
            #Normalization
        else:
            q_tilde = q_tilde / np.sqrt(sum(q_tilde**2))
            q.append(q_tilde)
    print('Vectors are linearly independent.')
    return q


### This is the QR factorization of A as it expresses martix A as the product of two matricies Q and R. 

In [24]:
A= np.random.rand(3,3)

In [25]:
def QR_factorization(A):
    #transpose A to use columns in gram schmidt
    Q_transpose = np.array(gram_schmidt(A.T))
    #create matrix R by multiplying Q by A
    R = Q_transpose @ A
    #transpose Q and return result
    Q = Q_transpose.T
    return Q, R

Q, R = QR_factorization(A)
Q,R

Vectors are linearly independent.


(array([[ 0.95167929, -0.21639852, -0.21789494],
        [ 0.23661592,  0.06442246,  0.96946514],
        [ 0.19575349,  0.97417731, -0.11251286]]),
 array([[ 5.57309046e-01,  1.11469757e+00,  1.12056523e+00],
        [ 2.40940253e-16,  7.39802466e-01,  4.61961492e-01],
        [-4.02054479e-16, -3.69729525e-15,  1.05980119e-01]]))

In [26]:
#compute Q and R factorization using numpy
Q, R = np.linalg.qr(A)
Q,R

(array([[-0.95167929,  0.21639852,  0.21789494],
        [-0.23661592, -0.06442246, -0.96946514],
        [-0.19575349, -0.97417731,  0.11251286]]),
 array([[-0.55730905, -1.11469757, -1.12056523],
        [ 0.        , -0.73980247, -0.46196149],
        [ 0.        ,  0.        , -0.10598012]]))

Show that $G = A^TA = (QR)^T(RQ) = R^TQ^TQR = R^T(Q^TQ) = R^TR$

In [27]:
G = A.T@A
G

array([[0.31059337, 0.62123104, 0.62450114],
       [0.62123104, 1.78985837, 1.59085159],
       [0.62450114, 1.59085159, 1.48030663]])

In [28]:
qrTrq = ( (Q@R).T) @ (Q@R) 
qrTrq

array([[0.31059337, 0.62123104, 0.62450114],
       [0.62123104, 1.78985837, 1.59085159],
       [0.62450114, 1.59085159, 1.48030663]])

In [29]:
rTqTqr = R.T @ (Q.T@Q) @ R
rTqTqr

array([[0.31059337, 0.62123104, 0.62450114],
       [0.62123104, 1.78985837, 1.59085159],
       [0.62450114, 1.59085159, 1.48030663]])

In [30]:
R.T@R

array([[0.31059337, 0.62123104, 0.62450114],
       [0.62123104, 1.78985837, 1.59085159],
       [0.62450114, 1.59085159, 1.48030663]])

In [31]:
print("Are all the intermediates above equivalent : ",G.all() == qrTrq.all() == rTqTqr.all() == (R.T@R).all())

Are all the intermediates above equivalent :  True


# 10.40 

The claim of this problem is that when you compute the QR factorization $A = QR$ you are also computing the QR factorization for all submatricies of the first $i$ colums of $A$. Here $A$ is a $n*k$ matrix which we run gram-schmidt on the columns of to obtain matricies $Q$ and $R$ (See solution 10.39 for an expanded form of the QR Factorization function not using numpy but the gram-schmidt algorithm). Now with this information the problem does make sense since we are using the gram-schmidt algorithm to compute $Q$ and $R$ given $A$. Since the columns of A are linerally independent (known to be true if gram-schmidt does not terminate) we can say any submatrix with column index less than k is also linerally independent. Here is an example in python. 

__Notes__

* $i$ is an index between 1 and $k$. 

$A_i = A_{1:n,1:i}$

$Q_i = Q_{1:n,1:i}$

$R_i = R_{1:i,1:i}$

In [32]:
A= np.random.rand(5,5)
print("Matrix A: \n", A,"\n \n","shape of A: ", A.shape,"\n")

#compute Q and R for matrix A
Q, R = np.linalg.qr(A)
print("Q : \n", Q)
print("R : \n", R)

Matrix A: 
 [[0.80311433 0.20958377 0.29296335 0.52555851 0.91898825]
 [0.33269765 0.99795833 0.33594934 0.56854303 0.90011599]
 [0.04822564 0.76421613 0.6599783  0.8270693  0.56290286]
 [0.15189166 0.77512675 0.10497224 0.27809627 0.8734102 ]
 [0.21128936 0.66607539 0.95667397 0.15360871 0.66646009]] 
 
 shape of A:  (5, 5) 

Q : 
 [[-0.88381401  0.4092056  -0.02986647  0.17352348 -0.14290262]
 [-0.36612826 -0.4912035   0.37114883 -0.07088541  0.69418516]
 [-0.05307152 -0.52048159 -0.30987867  0.7793914  -0.15101847]
 [-0.16715425 -0.45595544  0.46986744 -0.26252507 -0.6888173 ]
 [-0.23252044 -0.33537062 -0.73794061 -0.53711683 -0.02024745]]
R : 
 [[-0.90869155 -0.87561381 -0.65694452 -0.79875161 -1.47260653]
 [ 0.         -1.37900357 -0.75734681 -0.67299854 -0.98081283]
 [ 0.          0.         -0.74522131 -0.04365935  0.05077752]
 [ 0.          0.          0.          0.53999297 -0.0528764 ]
 [ 0.          0.          0.          0.         -0.20660157]]


Now lets look at all the sub matricies which start at the first column and remove the last column of the submatrix until we have run QR factorization on all sub matricies of the first i columns of A. This is, show that $A_i = Q_iR_i$ is the QR factorization of $A_i$ for each $i$

In [33]:
for i in range(len(A)-1):
    print ("A%s = \n" %(5-i),A[:,0:len(A)-1-i])
    #compute the submatricies to show they are equivalent to those computed wh
    Qsub, Rsub = np.linalg.qr(A[:,0:len(A)-1-i])
    print("Q%s = \n"%(5-i),Qsub,"\n","R%s = \n" %(5-i), Rsub)

A5 = 
 [[0.80311433 0.20958377 0.29296335 0.52555851]
 [0.33269765 0.99795833 0.33594934 0.56854303]
 [0.04822564 0.76421613 0.6599783  0.8270693 ]
 [0.15189166 0.77512675 0.10497224 0.27809627]
 [0.21128936 0.66607539 0.95667397 0.15360871]]
Q5 = 
 [[-0.88381401  0.4092056  -0.02986647  0.17352348]
 [-0.36612826 -0.4912035   0.37114883 -0.07088541]
 [-0.05307152 -0.52048159 -0.30987867  0.7793914 ]
 [-0.16715425 -0.45595544  0.46986744 -0.26252507]
 [-0.23252044 -0.33537062 -0.73794061 -0.53711683]] 
 R5 = 
 [[-0.90869155 -0.87561381 -0.65694452 -0.79875161]
 [ 0.         -1.37900357 -0.75734681 -0.67299854]
 [ 0.          0.         -0.74522131 -0.04365935]
 [ 0.          0.          0.          0.53999297]]
A4 = 
 [[0.80311433 0.20958377 0.29296335]
 [0.33269765 0.99795833 0.33594934]
 [0.04822564 0.76421613 0.6599783 ]
 [0.15189166 0.77512675 0.10497224]
 [0.21128936 0.66607539 0.95667397]]
Q4 = 
 [[-0.88381401  0.4092056  -0.02986647]
 [-0.36612826 -0.4912035   0.37114883]
 [-0.05

# 10.43
__Information needed:__ See pg. 182 of VMLS. 

Following from what is provided in problem 10.43. 

Lets assumse we are taking the inner product of the rows of $A$ and the columns of $B$. Then the complexity is ~ $2mnp$ flops. Most of the time we let $A$ be of size $m*p$ and $B$ be of size $p*n$. However here $A$ is the same size as $B$ so we know $m=p$ and $p=n$. This tells us all the terms are equivalent to $n$. Finally, since we know $m = p = n$ we have $2*n*n*n$ and by convention we can drop the constant to get the complexity of $n^3$ flops. 

Next, we are told it takes ~ 0.2 seconds to compute the product of two matricies each of size 1500*1500. This problem has a nice set up to solve with a ratio. This is, we have 3 knowns related to one another and one unknown related to the other 3. 

let $x1 = 0.2$ sec, $n1_{flops} = (1500)^3$ and $x2 = t$, $n2_{flops} = 3000^3$

$\frac{x1}{n1_{flops}} = \frac{x2}{n2_{flops}}$

Solve for the unkown $x_2$

$x_2 = \frac{x_1n2_{flops}}{n1_{flops}} =  1.6$ seconds

In practice this values will vary but it is a good back of the envelope approximation. 